libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,337 @@
+			stateVar->accept(builder);
+
+		for (auto const& inheritanceSpecifier: base->baseContracts())
+			inheritanceSpecifier->accept(builder);
+
+		if (base->constructor())
+		{
+			builder.functionReferenced(*base->constructor());
+
+			// Constructors and functions called in state variable initializers have an edge either from
+			// the previous class in linearized order or from Entry if there's no class before.
+			builder.m_currentNode = base->constructor();
+		}
+	}
+
+	builder.m_currentNode = SpecialNode::Entry;
+	builder.processQueue();
+
+	return make_unique<FunctionCallGraphBuilder::ContractCallGraph>(std::move(builder.m_graph));
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::buildDeploymentGraph(
+	ContractDefinition const& _contract,
+	FunctionCallGraphBuilder::ContractCallGraph const& _creationGraph
+)
+{
+	solAssert(&_creationGraph.contract == &_contract, "");
+
+	FunctionCallGraphBuilder builder(_contract);
+
+	auto getSecondElement = [](auto const& _tuple){ return get<1>(_tuple); };
+
+	// Create graph for all publicly reachable functions
+	for (FunctionTypePointer functionType: _contract.interfaceFunctionList() | views::transform(getSecondElement))
+	{
+		auto const* function = dynamic_cast<FunctionDefinition const*>(&functionType->declaration());
+		auto const* variable = dynamic_cast<VariableDeclaration const*>(&functionType->declaration());
+
+		if (function)
+			builder.functionReferenced(*function);
+		else
+			// If it's not a function, it must be a getter of a public variable; we ignore those
+			solAssert(variable, "");
+	}
+
+	if (_contract.fallbackFunction())
+		builder.functionReferenced(*_contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		builder.functionReferenced(*_contract.receiveFunction());
+
+	// All functions present in internal dispatch at creation time could potentially be pointers
+	// assigned to state variables and as such may be reachable at deployment time as well.
+	builder.m_currentNode = SpecialNode::InternalDispatch;
+	for (Node const& dispatchTarget: valueOrDefault(_creationGraph.edges, SpecialNode::InternalDispatch, {}))
+	{
+		solAssert(!holds_alternative<SpecialNode>(dispatchTarget), "");
+		solAssert(get<CallableDeclaration const*>(dispatchTarget) != nullptr, "");
+		solAssert(dynamic_cast<CallableDeclaration const*>(get<CallableDeclaration const*>(dispatchTarget)) != nullptr, "");
+
+		// Visit the callable to add not only it but also everything it calls too
+		builder.functionReferenced(*get<CallableDeclaration const*>(dispatchTarget), false);
+	}
+
+	builder.m_currentNode = SpecialNode::Entry;
+	builder.processQueue();
+
+	return make_unique<FunctionCallGraphBuilder::ContractCallGraph>(std::move(builder.m_graph));
+}
+
+bool FunctionCallGraphBuilder::visit(FunctionCall const& _functionCall)
+{
+	solAssert(holds_alternative<SpecialNode>(m_currentNode) || get<CallableDeclaration const*>(m_currentNode) != nullptr, "");
```

Comment:
Shouldn't this always be the case? If you really want to check this all the time, maybe change the type to a `std::reference_wrapper`?

Reply:
It should never be null if the code is correct but that's what assertions are for :)

Maybe `reference_wrapper` would be better than a pointer but I don't really want to change it everywhere at this point so I'll remove the assert. In this particular place I'm not dereferencing the pointer anyway, it was more of a "just in case" situation.